### PR TITLE
Add port scanning functionality

### PR DESCRIPTION
**Pull request by evansmurithi**
**NB:**

- You need to have `nmap` installed in the environment you are running this tool
- The scanning feature uses `SYN` and `UDP` scans which require root privileges
- Input format is as shown below:

  ```yaml
  port:
    - discovery:
        type: AWS
        resource_types:
          - EC2
        regions:
          - eu-central-1
        tags:
          App: U-Report
      allowlist:
        - 22
      timeout: 5m
    - discovery:
        type: host
        targets:
          - 197.210.136.181
          - 197.210.136.180
          - 197.210.136.178
          - 197.210.136.182
      allowlist:
        - 80
        - 443
      blocklist:
        - 22
      timeout: 5m
    - discovery:
        type: host
        targets:
          - 34.252.205.12
      allowlist:
        - 22
      timeout: 5m
  ```

### CODE REVIEW

File: libs/audit/port.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+		// check if port has "-" to denote a range
+		if strings.Contains(port, "-") {
+			a := strings.FieldsFunc(port, func(r rune) bool {
+				return r == '/' || r == '-'
+			})
+			protocol := a[0]
+			start, _ := strconv.Atoi(a[1])
+			end, _ := strconv.Atoi(a[2])
+
+			for i := start; i < end+1; i++ {
```
Comment (jasonrogena) [inline libs/audit/port.go:41]: @evansmurithi this piece of code here seems to add a time complexity of `O(n*m)` where `n` is the number of ports in the config range and `m` is the number of ports nmap returns. I wonder whether we can reduce the complexity by, let's say, looking up the ports using HashMaps (that give you a time complexity of `O(1)` during lookups).

https://blog.golang.org/maps

Comment (jasonrogena) [reply libs/audit/port.go:41]: Oh, seen where you've used a map. This looks good.

Comment (evansmurithi) [reply libs/audit/port.go:41]: I'm already using a HashMap to lookup the nmap ports (L26-28). The challenge I was facing was to flatten the user defined ranges and look them up in the HashMap

**Comment by jasonrogena** [comment]
@evansmurithi Thing's I've noticed so far:

1. Seems the nmap scan right now is scanning only popular ports, and not 1 – 65535. This means that a host can have an "obscure" port like tcp/13013 open and the scan result will not indicate this.
2. In the configuration file adding a port range e.g. 60000 - 61000, the scan fails unless all the ports in that range are listening. This might lead to issues especially for services like Mosh where you have a range of ports that are expected to be open but will only be listening when an SSH connection is active.

**Comment by evansmurithi** [comment]
To get the top X ports, I ran:

- TCP: sudo nmap -sS -oX - --top-ports 1000 x
- UDP: sudo nmap -sU -oX - --top-ports 1000 x

**Review by jasonrogena — APPROVED** [review]

**Pull request merged**